cpp/src/tsne/utils.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#define TEST_NNZ 12021
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include <cuda_runtime.h>
+#include <cuda_utils.h>
+
+/* Change to <sparse/coo.h> */
+#include "./sparse/coo.h"
+
+#include <thrust/device_ptr.h>
+#include <thrust/functional.h>
+#include <thrust/reduce.h>
+#include <thrust/transform.h>
+using namespace thrust::placeholders;
+
+#include <random/rng.h>
+#include <stats/sum.h>
+#include <sys/time.h>
+
+// ###################### General global funcs ######################
+#define thrust_t thrust::device_ptr
+#define to_thrust thrust::device_pointer_cast
+#define __STREAM__ thrust::cuda::par.on(stream)
+#define cuda_create_stream(x) CUDA_CHECK(cudaStreamCreate(x))
+#define cuda_destroy_stream(x) CUDA_CHECK(cudaStreamDestroy(x))
+#define cuda_synchronize() CUDA_CHECK(cudaDeviceSynchronize())
+
+#define COO_t Sparse::COO
+#define MAX(a, b) ((a > b) ? a : b)
```

Comment:
These functions already exist in c++ and cuda has variants that work in kernels as well. They should not need to be defined here.